### PR TITLE
Allows passing of instance variables to hash mappers

### PR DESCRIPTION
**Pull request by Berardpi**

### CODE REVIEW

**Comment by Berardpi** [comment]
@MarcoCode Can you take a look please?

File: app/models/maestrano/connector/rails/concerns/entity.rb [hunk 1]
```diff
@@ -94,16 +95,18 @@ def mapper_class
       raise 'Not implemented'
     end
 
+    # Optional creation only mapper, default to main mapper
```
Comment (MarcoCode) [inline app/models/maestrano/connector/rails/concerns/entity.rb]: Being annoying: 'defaults' to

File: app/models/maestrano/connector/rails/concerns/entity.rb [hunk 2]
```diff
@@ -226,15 +246,22 @@ def get_connec_entities(last_synchronization_date = nil)
...
+  # Pushes the external entities to Connec!, and updates the idmaps with either
+  # * connec_id and push timestamp
+  # * error message
   def push_entities_to_connec_to(mapped_external_entities_with_idmaps, connec_entity_name)
     return unless self.class.can_write_connec?
 
     Maestrano::Connector::Rails::ConnectorLogger.log('info', @organization, "Sending #{Maestrano::Connector::Rails::External.external_name} #{self.class.external_entity_name.pluralize} to Connec! #{connec_entity_name.pluralize}")
 
+    # As we're doing only push, we use the idmaps to filter out updates
     unless self.class.can_update_connec?
```
Comment (MarcoCode) [inline app/models/maestrano/connector/rails/concerns/entity.rb]: Did you mean post?

File: app/models/maestrano/connector/rails/concerns/entity.rb [hunk 3]
```diff
@@ -160,14 +167,24 @@ def public_external_entity_name
...
   def map_to_external(entity, first_time_mapped = nil)
     mapper = first_time_mapped ? self.class.creation_mapper_class : self.class.mapper_class
-    mapper.normalize(entity).with_indifferent_access
+    map_to_external_helper(entity, mapper)
+  end
+
+  def map_to_external_helper(entity, mapper)
+    # instance_values returns a hash with all the instance variables (http://apidock.com/rails/v4.0.2/Object/instance_values)
+    # that includes opts, organization, connec_client, external_client, and all the connector specific variables
+    mapper.normalize(entity, instance_values.with_indifferent_access).with_indifferent_access
```
Comment (MarcoCode) [inline app/models/maestrano/connector/rails/concerns/entity.rb]: Nice! 👍

File: app/models/maestrano/connector/rails/concerns/entity.rb [hunk 4]
```diff
@@ -160,14 +167,24 @@ def public_external_entity_name
   # Map a Connec! entity to the external model
   def map_to_external(entity, first_time_mapped = nil)
     mapper = first_time_mapped ? self.class.creation_mapper_class : self.class.mapper_class
-    mapper.normalize(entity).with_indifferent_access
+    map_to_external_helper(entity, mapper)
+  end
+
+  def map_to_external_helper(entity, mapper)
+    # instance_values returns a hash with all the instance variables (http://apidock.com/rails/v4.0.2/Object/instance_values)
+    # that includes opts, organization, connec_client, external_client, and all the connector specific variables
```
Comment (MarcoCode) [inline app/models/maestrano/connector/rails/concerns/entity.rb]: Did you mean 'entity' specific variables?

**Comment by MarcoCode** [comment]
👍

**Pull request merged**